### PR TITLE
6923/feature/implementing fuzzy search

### LINKED ISSUE #6923: Add support for solr spell checking ("Did you mean?")
Currently, search is too strict with letters. Compare [Безпека життєдіяльності](https://openlibrary.org/search?q=%D0%91%D0%B5%D0%B7%D0%BF%D0%B5%D0%BA%D0%B0+%D0%B6%D0%B8%D1%82%D1%82%D1%94%D0%B4%D1%96%D1%8F%D0%BB%D1%8C%D0%BD%D0%BE%D1%81%D1%82%D1%96&mode=everything) and [Безпека життєдіяльност](https://openlibrary.org/search?q=%D0%91%D0%B5%D0%B7%D0%BF%D0%B5%D0%BA%D0%B0+%D0%B6%D0%B8%D1%82%D1%82%D1%94%D0%B4%D1%96%D1%8F%D0%BB%D1%8C%D0%BD%D0%BE%D1%81%D1%82&mode=everything). With just one letter missing ( **_і_** ) there are no results.
This possibly can be solved by solr spell checking feature
https://solr.apache.org/guide/6_6/spell-checking.html
### Stakeholders
@cdrini

### PR DESCRIPTION
**Pull request by benbdeitch**
Closes #6923 

This PR implements fuzzy searching on Solr for fieldless searches, as well as for searches for the 'title', 'alternative_title', 'author', and 'text' fields.  Notably, this is currently only active for work searches (and related edition queries), but it can easily be transferred to author searches, etc, if requested.  

### Technical

Queries of this sort will match single word terms that can be transformed into one another by up to two operations of insertion, substitution, or deletion. This sort of functionality is built into Solr, and as such, editing the way in which we write queries was the only change needed. 

Unfortunately, there are some limitations to fuzzy searching; it does not function with phrases, or a series of words wrapped in quotation marks. . Fortunately, this preserves the expected functionality of many search engines, as a means of requesting only  exact matches.  

In order to ease editing, a new constant in the Work SearchScheme was introduced to control which fields will be fuzzy when queried. 
### Testing

Simply search on Solr, under the books tag. Terms without a field will be affected, as will terms related to the fields mentioned 
### Screenshot

image

### Stakeholders

@cdrini

### CODE REVIEW

**Comment by natedunn** [comment]
Thanks so much for working on this. This would honestly save me such a headache if this got merged in soon. Wish I could contribute more, but I am just a mere consumer of the JSON API.

**Comment by benbdeitch** [comment]
It was honestly my pleasure, and I'm glad to hear that it'll help you out! I think there's currently some Solr performance issues being investigated, but I imagine it'll get merged in relatively soon.

File: openlibrary/solr/query_utils.py [hunk 1]
```diff
@@ -279,3 +278,62 @@ def luqum_replace_field(query, replacer: Callable[[str], str]) -> str:
...
+        for child in item.children:
+            yield from luqum_field_traverse(child, new_parents)
+
+
+def luqum_make_fuzzy(tree: Item, search_fields: set[str] | None = None):
+    """:param query: Passed in the form of a luqum tree
+    :param search_fields: A list of search fields to be fuzzied out."""
+
+    if not search_fields:
+        search_fields = set({})
```
Comment (cdrini) [inline openlibrary/solr/query_utils.py:307]:
```suggestion
        search_fields = set()
```

File: openlibrary/solr/query_utils.py [hunk 2]
```diff
@@ -279,3 +278,62 @@ def luqum_replace_field(query, replacer: Callable[[str], str]) -> str:
...
+    Nuch like its sister function, it does not make any guarantees about what will happen if you
+    modify the tree while traversing it 😅 But we do it anyways.
+
+    :param item: Node to traverse
+    :param _parents: Internal parameter for tracking parents
+    """
+    parents = _parents or []
+    yield item, parents
+    new_parents = [*parents, item]
+    if not isinstance(item, SearchField):
```
Comment (cdrini) [inline openlibrary/solr/query_utils.py:297]: Since this is the only difference between this and `luqum_traverse`, adding a new parameter to `luqum_traverse` , maybe `stop_at: Tuple[Class] = None` which we can set to SearchField might be a useful way to DRY these.

File: openlibrary/solr/query_utils.py [hunk 3]
```diff
@@ -279,3 +278,62 @@ def luqum_replace_field(query, replacer: Callable[[str], str]) -> str:
...
+        for child in item.children:
+            yield from luqum_field_traverse(child, new_parents)
+
+
+def luqum_make_fuzzy(tree: Item, search_fields: set[str] | None = None):
+    """:param query: Passed in the form of a luqum tree
+    :param search_fields: A list of search fields to be fuzzied out."""
```
Comment (cdrini) [inline openlibrary/solr/query_utils.py:304]: Let's add some more tests in this doc string testing some edge cases eg

- parentheses
- quotations
- greedy fields, eg `lord of the rings author:j r r tolkein`

**Review by cdrini — CHANGES_REQUESTED** [review]
Looks good! A few code changes. Once #8821 is merged, I'll put this up on testing and A/B compare against prod using https://docs.google.com/spreadsheets/d/1BN5I7-OkTPaoTr2Es6jQ4O9ICWFmH0q9CP6kEgolCgg/edit#gid=1006480604 to see how we're doing!

(Note to self: Depending on results of testing, we might want to only fuzz strings that are greater than length 2)

File: openlibrary/solr/query_utils.py (hunk 3 above)
Comment (benbdeitch) [reply openlibrary/solr/query_utils.py:304]: It looks like the tests were already largely included; I added one additional one for the parentheses, but all of the other situations seemed to be covered.

File: openlibrary/solr/query_utils.py [hunk 4]
```diff
@@ -279,3 +278,62 @@ def luqum_replace_field(query, replacer: Callable[[str], str]) -> str:
         if isinstance(sf, SearchField):
             sf.name = replacer(sf.name)
     return str(query)
+
+
+def luqum_field_traverse(item: Item, _parents: list[Item] | None = None):
```
[the remark this answers is not in the document]
Comment (benbdeitch) [reply openlibrary/solr/query_utils.py]: Done!

File: openlibrary/solr/query_utils.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (benbdeitch) [reply openlibrary/solr/query_utils.py]: Done

File: openlibrary/solr/query_utils.py [hunk 5]
```diff
@@ -279,3 +278,62 @@ def luqum_replace_field(query, replacer: Callable[[str], str]) -> str:
...
+    """
+    parents = _parents or []
+    yield item, parents
+    new_parents = [*parents, item]
+    if not isinstance(item, SearchField):
+        for child in item.children:
+            yield from luqum_field_traverse(child, new_parents)
+
+
+def luqum_make_fuzzy(tree: Item, search_fields: set[str] | None = None):
```
[the remark this answers is not in the document]
Comment (benbdeitch) [reply openlibrary/solr/query_utils.py]: Done

File: openlibrary/solr/query_utils.py (hunk 1 above)
Comment (benbdeitch) [reply openlibrary/solr/query_utils.py]: Done

File: openlibrary/solr/query_utils.py (hunk 2 above)
Comment (benbdeitch) [reply openlibrary/solr/query_utils.py]: Done, but changed to a Set from a tuple, so that the check for inclusion is quicker.

File: openlibrary/solr/query_utils.py (hunk 3 above)
Comment (benbdeitch) [reply openlibrary/solr/query_utils.py]: Done

**Comment by cdrini** [comment]
Ah shoot I put this on testing but it's causing a lot of timeouts; likely because of #5480 , since the search result sets are much larger now. I think this is blocked until that issue is resolved.

**Comment by RomanKlimov** [comment]
what's the status of this one?

**Comment by tfmorris** [comment]
I followed a link here from a blog post which said this was done, but it seems to have been abandoned a couple of years ago. What's the actual status?

This would be much better than the current fallback to full text search which is almost never what the user wants when they've simply made a typo.

**Comment by cdrini** [comment]
Ah that was an error in the post; I corrected it. My last comment here likely still stands; when we last tested it, it had a significant performance impact, so I'm not sure if fuzzy search is a tenable solution to this problem. Closing this PR for now, since there isn't really a step forward with this approach. Our best bet would likely be trying to evaluate solr's "Did you mean" support to compare its performance, or to investigate solutions to #5480 .
